portable/GCC/Posix/port.c

```diff
@@ -0,0 +1,605 @@
+	{
+		vPortEnableInterrupts();
+	}
+}
+/*-----------------------------------------------------------*/
+
+void vPortYieldFromISR( void )
+{
+Thread_t *xThreadToSuspend;
+Thread_t *xThreadToResume;
+
+	xThreadToSuspend = prvGetThreadFromTask( xTaskGetCurrentTaskHandle() );
+
+	vTaskSwitchContext();
+
+	xThreadToResume = prvGetThreadFromTask( xTaskGetCurrentTaskHandle() );
+
+	prvSwitchThread( xThreadToResume, xThreadToSuspend );
+}
+/*-----------------------------------------------------------*/
+
+void vPortYield( void )
+{
+	vPortEnterCritical();
+
+	vPortYieldFromISR();
+
+	vPortExitCritical();
+}
+/*-----------------------------------------------------------*/
+
+void vPortDisableInterrupts( void )
+{
+	pthread_sigmask( SIG_BLOCK, &xAllSignals, NULL );
+}
+/*-----------------------------------------------------------*/
+
+void vPortEnableInterrupts( void )
+{
+	pthread_sigmask( SIG_UNBLOCK, &xAllSignals, NULL );
+}
+/*-----------------------------------------------------------*/
+
+portBASE_TYPE xPortSetInterruptMask( void )
+{
+	/* Interrupts are always disabled inside ISRs (signals
+	   handlers). */
+	return pdTRUE;
+}
+/*-----------------------------------------------------------*/
+
+void vPortClearInterruptMask( portBASE_TYPE xMask )
+{
+}
+/*-----------------------------------------------------------*/
+
+static uint64_t prvGetTimeNs(void)
+{
+struct timespec t;
+
+	clock_gettime(CLOCK_MONOTONIC, &t);
+
+	return t.tv_sec * 1000000000ull + t.tv_nsec;
+}
+
+static uint64_t prvStartTimeNs;
+static uint64_t prvTickCount;
+
+/*
+ * Setup the systick timer to generate the tick interrupts at the required
+ * frequency.
+ */
+void prvSetupTimerInterrupt( void )
+{
+struct itimerval itimer;
+int iRet;
+
+	/* Initialise the structure with the current timer information. */
+	iRet = getitimer( ITIMER_REAL, &itimer );
+	if ( iRet )
+	{
+		prvFatalError( "getitimer", errno );
+	}
+
+	/* Set the interval between timer events. */
+	itimer.it_interval.tv_sec = 0;
+	itimer.it_interval.tv_usec = portTICK_RATE_MICROSECONDS;
+
+	/* Set the current count-down. */
+	itimer.it_value.tv_sec = 0;
+	itimer.it_value.tv_usec = portTICK_RATE_MICROSECONDS;
+
+	/* Set-up the timer interrupt. */
+	iRet = setitimer( ITIMER_REAL, &itimer, NULL );
+	if ( iRet )
+	{
+		prvFatalError( "setitimer", errno );
+	}
+
+	prvStartTimeNs = prvGetTimeNs();
+}
+/*-----------------------------------------------------------*/
+
+static void vPortSystemTickHandler( int sig )
+{
+Thread_t *pxThreadToSuspend;
+Thread_t *pxThreadToResume;
+uint64_t xExpectedTicks;
+
+	uxCriticalNesting++; /* Signals are blocked in this signal handler. */
+
+	pxThreadToSuspend = prvGetThreadFromTask( xTaskGetCurrentTaskHandle() );
+
+	/* Tick Increment, accounting for any lost signals or drift in
+	 * the timer. */
+	xExpectedTicks = (prvGetTimeNs() - prvStartTimeNs)
+		/ (portTICK_RATE_MICROSECONDS * 1000);
+	do {
+		xTaskIncrementTick();
+		prvTickCount++;
+	} while (prvTickCount < xExpectedTicks);
```

Comment:
The timer list is ordered so timers will still be called in the expected order and blocked tasks will still be woken in the expected order.

I think some small amount of jitter in the tick is much better than trying to deal with lost ticks or tick drift.

Also, since this is an application running on a non-RTOS there should be no expectation of real-time behaviour precisely matching the behaviour on a real bare-metal target.